scripts/field_report.py

```diff
@@ -0,0 +1,262 @@
+        # pylint: disable=redefined-builtin
+        """Get the base name of a type, without typedefs, qualifiers or pointers."""
+        core = type
+        lower = type # type: Optional[clang.cindex.Type]
+        while lower:
+            core, lower = lower, self.get_underlying_type(core)
+        # There's no API function to remove qualifiers from a type,
+        # so do it textually. Remove 'const', 'restrict', etc.
+        # Also remove 'struct', so we'll get the struct name from struct
+        # definitions.
+        return re.sub(self.QUALIFIERS_RE, r'', core.spelling)
+
+    def record_field_access(self, node: Cursor) -> None:
+        """Record one location where a field is accessed."""
+        field_name = node.spelling
+        lhs = next(node.get_children())
+        structure_type = self.get_type_core(lhs.type)
+        if structure_type not in self.fields:
+            # This is not a structure defined by the library
+            return
+        self.fields[structure_type][field_name].record_use(node)
+
+    def sanity_checks(self) -> None:
+        """If the data looks wrong, raise a `SanityCheck` exception."""
+        for type_name in sorted(self.fields):
+            for field in self.fields[type_name].values():
+                if field.offset == -1:
+                    raise SanityCheck('Could not determine offset of {}.{}. '
+                                      .format(type_name, field))
+
+    def run_analysis(self, files: List[str],
+                     sanity_checks: bool = True) -> None:
+        self.read_field_definitions(files)
+        self.read_field_usage(files)
+        if sanity_checks:
+            self.sanity_checks()
+
+    def read_field_usage(self, filenames: List[str]) -> None:
+        """Parse field usage in the given C source files."""
+        self.load(*filenames)
+        for filename in filenames:
+            for node in self.files[filename].cursor.walk_preorder():
+                if not self.in_interesting_file(node.location):
+                    continue
+                if node.kind == CursorKind.MEMBER_REF_EXPR:
+                    self.record_field_access(node)
+
+    @staticmethod
+    def report_field(out: typing_util.Writable,
+                     prefix: str, field: FieldInfo) -> None:
+        """Print information about a structure field.
+
+        Format: <type>.<name>,<size>,<alignment>,<offset>,<use_count>,<score>
+        """
+        out.write('{},{},{},{},{},{}\n'.format(
+            prefix + field.node.spelling,
+            field.size, field.align, field.offset,
+            field.uses(), field.score()
+        ))
+
+    def report_fields(self, out: typing_util.Writable,
+                      header=False) -> None:
+        """Print information about fields of structures defined by the library."""
+        if header:
+            out.write('field,size,align,offset,uses,score\n')
+        for type_name in sorted(self.fields):
```

Comment:
`self.fields` is created as an `OrderedDict()`.  Yet we sort through it each time we iterate. It feels a bit redundant. 

Same to line 183

Reply:
`OrderedDict` because the order of the fields in the structure is semantically significant and it seemed useful to preserve it. (Granted we can sort by offset.) `sorted` when I want a list of field names in a canonical form. I think that's reasonable, even if we're not using the field order right now.